### PR TITLE
Add support for Python 3.9

### LINKED ISSUE #847: Please provide python 3.9 support
Vdirsyncer emits error on MacOS Catalina as well as on Void Linux being compiled against Python 3.9.

Example -vdebug output on Linux:

debug: PROPFIND https://carddav.yandex.ru/addressbook/myname%40yandex.ru/addressbook/
debug: {'User-Agent': 'vdirsyncer/0.16.8', 'Content-Type': 'application/xml; charset=UTF-8', 'Depth': '1'}
debug: b'<?xml version="1.0" encoding="utf-8" ?>\n            <D:propfind xmlns:D="DAV:">\n                <D:prop>\n                    <D:resourcetype/>\n                    <D:getcontenttype/>\n                    <D:getetag/>\n   
...
<D:propstat><D:prop><D:getetag>"2495-2727"</D:getetag><D:getcontenttype>text/x-vcard</D:getcontenttype><D:resourcetype/></D:prop><status xmlns="DAV:">HTTP/1.1 200 OK</status></D:propstat></D:response></D:multistatus>'
debug: Normalized URL from '/addressbook/myname%40yandex.ru/addressbook/' to '/addressbook/myname@yandex.ru/addressbook/'
error: Unknown error occurred for avg_contacts/addressbook: 'xml.etree.ElementTree.Element' object has no attribute 'getiterator'
error: Use `-vdebug` to see the full traceback.
debug:   File "/usr/lib/python3.9/site-packages/vdirsyncer/cli/tasks.py", line 62, in sync_collection
debug:     sync.sync(
debug:   File "/usr/lib/python3.9/site-packages/vdirsyncer/sync/__init__.py", line 135, in sync
debug:     b_nonempty = b_info.prepare_new_status()
debug:   File "/usr/lib/python3.9/site-packages/vdirsyncer/sync/__init__.py", line 44, in prepare_new_status
debug:     for href, etag in self.storage.list():
debug:   File "/usr/lib/python3.9/site-packages/vdirsyncer/storage/dav.py", line 635, in list
debug:     for href, etag, _prop in rv:
debug:   File "/usr/lib/python3.9/site-packages/vdirsyncer/storage/dav.py", line 591, in _parse_prop_responses
debug:     props = _merge_xml(props)
debug:   File "/usr/lib/python3.9/site-packages/vdirsyncer/storage/dav.py", line 123, in _merge_xml
debug:     rv.extend(item.getiterator())
error: 1 out of 2 tasks failed.

My config:

[general]
status_path = "~/.vdirsyncer/status/"

[pair avg_contacts]
a = "avg_contacts_local"
b = "avg_contacts_yandex"
collections = ["from a", "from b"]
metadata = ["displayname"]
conflict_resolution = "a wins"

[storage avg_contacts_local]
type = "filesystem"
path = "~/Documents/Contacts/"
fileext = ".vcf"

[storage avg_contacts_yandex]
type = "carddav"
url = "https://carddav.yandex.ru/"
username = "myname@yandex.ru"
password.fetch = ["command", "keyring", "get", "yandex", "myname@yandex.ru"]

### LINKED ISSUE #854: Test issues with python3.9
Before you submit bug reports: https://vdirsyncer.pimutils.org/en/stable/contributing.html

Things to include in your bugreport:

* Your vdirsyncer version: 0.16.8
* If applicable, which server software (and which version) you're using: n/a
* Your Python version: 3.9.0
* Your operating system: Arch Linux
* Your config file: n/a
* Use `vdirsyncer -vdebug` for debug output. The output is sensitive, but
  please attach at least the last few lines before the error (if applicable),
  censored as necessary. This is almost always the most useful information.

When trying to build against python 3.9.0 and running the test suite I get:

```
py.test
============================= test session starts ==============================
platform linux -- Python 3.9.0, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /build/vdirsyncer/src/vdirsyncer, configfile: setup.cfg
plugins: subtesthack-0.1.1, localserver-0.5.0, hypothesis-5.41.2
collected 532 items

tests/storage/test_filesystem.py ....................................... [  7%]
.......sss......ss.......                                                [ 12%]
tests/storage/test_http.py .....                                         [ 12%]
tests/storage/test_http_with_singlefile.py ............................. [ 18%]
.........sssssss....sssss.ss                                             [ 23%]
tests/storage/test_memory.py ......................................sssss [ 31%]
ss....sss...ss                                                           [ 34%]
tests/storage/test_singlefile.py ....................................... [ 41%]
.............ss.ss                                                       [ 45%]
tests/storage/dav/test_caldav.py sssssssssssssssssssssssssssssssssssssss [ 52%]
sssssssssssssss                                                          [ 55%]
tests/storage/dav/test_carddav.py ssssssssssssssssssssssssss             [ 60%]
tests/storage/dav/test_main.py F................................         [ 66%]
tests/storage/etesync/test_main.py sssssssssssssssssssssssssssssssssssss [ 73%]
sssssssss                                                                [ 75%]
tests/system/cli/test_config.py .........                                [ 76%]
tests/system/cli/test_discover.py .....                                  [ 77%]
tests/system/cli/test_fetchparams.py .                                   [ 77%]
tests/system/cli/test_repair.py .....                                    [ 78%]
tests/system/cli/test_sync.py ........F...........                       [ 82%]
tests/system/cli/test_utils.py ..                                        [ 82%]
tests/system/utils/test_main.py .....                                    [ 83%]
tests/unit/test_exceptions.py .                                          [ 84%]
tests/unit/test_metasync.py .........                                    [ 85%]
tests/unit/test_repair.py ..........                                     [ 87%]
tests/unit/cli/test_config.py .                                          [ 87%]
tests/unit/cli/test_discover.py ......                                   [ 88%]
tests/unit/cli/test_fetchparams.py .........                             [ 90%]
tests/unit/sync/test_status.py .                                         [ 90%]
tests/unit/sync/test_sync.py ...............................             [ 96%]
tests/unit/utils/test_vobject.py ..................                      [100%]

=================================== FAILURES ===================================
______________________________ test_xml_utilities ______________________________
tests/storage/dav/test_main.py:28: in test_xml_utilities
    props = _merge_xml(response.findall('{DAV:}propstat/{DAV:}prop'))
vdirsyncer/storage/dav.py:123: in _merge_xml
    rv.extend(item.getiterator())
E   AttributeError: 'xml.etree.ElementTree.Element' object has no attribute 'getiterator'
___________________________ test_create_collections ____________________________
tests/system/cli/test_sync.py:273: in test_create_collections
    ???
E   hypothesis.errors.MultipleFailures: Hypothesis found 3 failures in explicit examples.
---------------------------------- Hypothesis ----------------------------------
Falsifying explicit example: test_create_collections(
    subtest=<function pytest_subtesthack.subtest.<locals>.inner>,
    collections=['persönlich'],
)
Traceback (most recent call last):
  File "/build/vdirsyncer/src/vdirsyncer/tests/system/cli/test_sync.py", line 294, in test_create_collections
    def test_inner(tmpdir, runner):
  File "/usr/lib/python3.9/site-packages/pytest_subtesthack.py", line 9, in inner
    item = Function(
  File "/usr/lib/python3.9/site-packages/_pytest/nodes.py", line 105, in __call__
    fail(msg, pytrace=False)
  File "/usr/lib/python3.9/site-packages/_pytest/outcomes.py", line 153, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: Direct construction of Function has been deprecated, please use Function.from_parent.
See https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent for more details.

Falsifying explicit example: test_create_collections(
    subtest=<function pytest_subtesthack.subtest.<locals>.inner>,
    collections={'A', 'a'},
)
Traceback (most recent call last):
  File "/build/vdirsyncer/src/vdirsyncer/tests/system/cli/test_sync.py", line 294, in test_create_collections
    def test_inner(tmpdir, runner):
  File "/usr/lib/python3.9/site-packages/pytest_subtesthack.py", line 9, in inner
    item = Function(
  File "/usr/lib/python3.9/site-packages/_pytest/nodes.py", line 105, in __call__
    fail(msg, pytrace=False)
  File "/usr/lib/python3.9/site-packages/_pytest/outcomes.py", line 153, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: Direct construction of Function has been deprecated, please use Function.from_parent.
See https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent for more details.

Falsifying explicit example: test_create_collections(
    subtest=<function pytest_subtesthack.subtest.<locals>.inner>,
    collections={'\ufffe'},
)
Traceback (most recent call last):
  File "/build/vdirsyncer/src/vdirsyncer/tests/system/cli/test_sync.py", line 294, in test_create_collections
    def test_inner(tmpdir, runner):
  File "/usr/lib/python3.9/site-packages/pytest_subtesthack.py", line 9, in inner
    item = Function(
  File "/usr/lib/python3.9/site-packages/_pytest/nodes.py", line 105, in __call__
    fail(msg, pytrace=False)
  File "/usr/lib/python3.9/site-packages/_pytest/outcomes.py", line 153, in fail
    raise Failed(msg=msg, pytrace=pytrace)
Failed: Direct construction of Function has been deprecated, please use Function.from_parent.
See https://docs.pytest.org/en/stable/deprecations.html#node-construction-changed-to-node-from-parent for more details.
```

Especially `test_xml_utilities` might be a real issue, while the rest is probably due to https://github.com/pimutils/vdirsyncer/pull/851

### PR DESCRIPTION
**Pull request by WhyNotHugo**
Fixes #847
Fixes #854

Also replaced Travis CI with [sourcehut's offering](https://builds.sr.ht).

`ETESYNC_TESTS` are currently not running, I'll address that soon separately.

Only tests with Python 3.9 are running now. I'll evaluate how important it is to run _the whole suite_ with older Pythons. I feel it might be an overkill.

However, we do have a build that runs with older dependencies. It might make sense to replace that with an Debian-based one, and actually install dependencies from Debian's repositories. Debian makes sense as a reference of "what's the oldest we're willing to support". Anyone with older dependencies around can just use an older version of vdirsyncer.

### CODE REVIEW

**Comment by WhyNotHugo** [comment]
Depends on https://github.com/untitaker/pytest-subtesthack/pull/5 being released.

**Comment by Tonus1** [comment]
Is there any news on python3.9 support ? I unfirtubately upgraded a bit too fast...

**Comment by WhyNotHugo** [comment]
The above issue is blocking upgrading. I might look into possibly dropping the dependency is that ends up being feasable.

**Comment by dvzrv** [comment]
@WhyNotHugo we're currently in the process of rebuilding all things on Arch against python3.9.
Unfortunately the test suite for vdirsyncer is still broken.

I'll probably disable the tests for the time being if this is further blocked by pytest-subtesthack.

File: tests/system/cli/test_sync.py [hunk 1]
```diff
@@ -271,6 +271,9 @@ def get_cfg():
     }
 
 
+hack = 0
```
Comment (dvzrv) [inline tests/system/cli/test_sync.py:274]: :O

**Comment by WhyNotHugo** [comment]
`httpsserver` from `pytest-localserver` seems to fail since it creates what is now considered insecure certificates.

The issue is that the project has vanished; there's not website for it, or anything online aside from the PyPI tarballs.

**Comment by d7415** [comment]
> The issue is that the project has vanished; there's not website for it, or anything online aside from the PyPI tarballs.

[Looks like](https://github.com/pytest-dev/pytest/issues/7857#issuecomment-704971024) they didn't migrate it when they moved everything else. There's a copy [here](https://bitbucket-archive.softwareheritage.org/projects/py/pytest-dev/pytest-localserver.html) but it doesn't sound (in the linked issue) like they're planning to maintain it :disappointed:

**Comment by fourstepper** [comment]
May I ask at what state this is right now? Is the plan to remove the dependency and/or is there a way to patch around for now?

**Comment by WhyNotHugo** [comment]
Fedora's patch seems unrelated to the above issue.

The most likely course of action is to drop this dependency. What is does it run a local HTTPS server, and the test checks that vdirsyncer validates it properly. I think the simplest thing to do would be to just run the simplest HTTPS server is come docker container (e.g.: haproxy? nginx?) and use that instead.

This kinda aligns with how we run CalDAV servers for tests too.

The main issue with the current dependency, is that is generates certs that are too insecure by today's standards. Creating a fork of that lib and patching it seems to much burden in the long run.

**Comment by WhyNotHugo** [comment]
Travis is no longer free and we're out of credits.
Gonna have to move CI onto something else :disappointed:

**Comment by d7415** [comment]
> Travis is no longer free and we're out of credits.

I guess that probably applies to all of pimutils? You can apply for "OSS credits" but you need to keep applying for them (so it's fine short term, but a pain longer term).

**Comment by geier** [comment]
I have requested some OSS credits, let's see if this helps us.

**Comment by WhyNotHugo** [comment]
Oh, neat, thanks!

**Comment by fourstepper** [comment]
I would be happy to try migrate the CI for this (and/or other projects) to an alternative CI system. Sourcehut build system (https://man.sr.ht/builds.sr.ht/) and GH Actions come to mind.

**Comment by WhyNotHugo** [comment]
@geier Have you gotten any reply from Travis? OSS credits still seem to be 0.

@fourstepper I really think fixing everything on Travis first makes sense the most -- it'd be really hard to migrate CI right now since many tests are failing due to non-CI reasons (e.g.: broken dependencies, etc).

However, if you want to give GitHub Actions a shot, you're more than welcome. Be warned that the CI setup for this repo is pretty non-trivial (but let me know if you've questions).

**Comment by geier** [comment]
nope, only got a generic "Thank you for your support request, we'll get back to you shortly" mail. The good thing about github actions would be, that we could set up our own runner. But I definitively don't have the time to port vdirsyncer. I'm trying to do this for khal at the moment (see github actions branch). Feel free to have a look everyone, help is appreciated.

**Comment by geier** [comment]
I got a reply from travis on the weekend, we have 25k new credits. When they are used up, we can re-apply for more credits.

**Comment by WhyNotHugo** [comment]
Thanks for the followup.

I've managed to set up CI on https://builds.sr.ht/ though, but I'll keep the credits in mind if need comes.

**Comment by geier** [comment]
@WhyNotHugo great! travis' loss.

**Pull request merged**